python/workflows.py

```diff
@@ -1,53 +1,88 @@
-       self.compensations = []
-
-
-   def add(self, function: typing.Awaitable):
-       self.compensations.append(function)
-
-
-   def __iadd__(self, function: typing.Awaitable):
-       self.add(function)
-       return self
-  
-   async def compensate(self):
-       if self.parallel_compensations:
-           try:
-               await asyncio.gather(self.compensations, return_exceptions=True)
-           except Exception as e:
-               print('failed to compensate: %s' % e)
-       else:
+    def __init__(self, parallel_compensations=False):
+        self.parallel_compensations = parallel_compensations
+        self.compensations = []
+
+    def add(self, function: typing.Callable[..., typing.Awaitable[None]]):
+        self.compensations.append(function)
+
+    def __iadd__(self, function: typing.Callable[..., typing.Awaitable[None]]):
+        self.add(function)
+        return self
+
+    async def compensate(self):
+        if self.parallel_compensations:
+
+            def compensation_lambda(f):
+                return workflow.execute_activity(
+                    f,
+                    start_to_close_timeout=time_delta,
+                    retry_policy=common_retry_policy,
+                )
+
+            all_compensations = [compensation_lambda(c) for c in self.compensations]
+            results = await asyncio.gather(*all_compensations, return_exceptions=True)
+            for result in results:
+                if isinstance(result, Exception):
+                    print("failed to compensate: %s" % e)
+
+        else:
             for f in reversed(self.compensations):
-               try:
-                    await f()
-               except Exception as e:
-                    print('failed to compensate: %s' % e)
+                try:
+                    await workflow.execute_activity(
+                        f,
+                        start_to_close_timeout=time_delta,
+                        retry_policy=common_retry_policy,
+                    )
+                except Exception as e:
+                    print("failed to compensate: %s" % e)
+
 
 @workflow.defn
 class BreakfastWorkflow:
     @workflow.run
-    async def run(self) -> None:
-       compensations = Compensations()
-       try:
-           await workflow.execute_activity(get_bowl, start_to_close_timeout=timedelta(seconds=5))
-           compensations += put_bowl_away
-           await workflow.execute_activity(add_cereal, start_to_close_timeout=timedelta(seconds=5))
-           compensations += put_cereal_back_in_box
-           await workflow.execute_activity(add_milk, start_to_close_timeout=timedelta(seconds=5))
-       except (ActivityError, CancelledError):
-           task = asyncio.create_task(compensations.compensate())
-           # Ensure the compensations run in the face of cancelation.
-           await asyncio.shield(task)
-           raise
+    async def run(self, parallel_compensations=False) -> None:
```

Comment:
I don't think we should encourage default parameter values in workflows (though they are technically allowed)